pkg/pillar/cmd/zedagent/attesttask.go

```diff
@@ -24,19 +32,197 @@ type WatchdogImpl struct{}
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected type from opaque ctx: %T",
+			ctx.OpaqueCtx)
+	}
+	if len(attestCtx.Nonce) > 0 {
+		//Clear existing nonce before attempting another nonce request
+		unpublishAttestNonce(attestCtx)
+		attestCtx.Nonce = nil
+	}
+	var attestReq = &attest.ZAttestReq{}
+
+	// bail if V2API is not supported
+	if !zedcloud.UseV2API() {
+		return zattest.ErrControllerReqFailed
+	}
+
+	attestReq.ReqType = attest.ZAttestReqType_ATTEST_REQ_NONCE
+	log.Debugf("Sending Nonce request %v", attestReq)
+	_, contents, senderStatus, err := trySendToController(attestReq, attestCtx.Iteration)
+	if err != nil || senderStatus != types.SenderStatusNone {
+		log.Errorf("[ATTEST] Error %v, senderStatus %v",
+			err, senderStatus)
+		//Increment Iteration for interface rotation next time
+		attestCtx.Iteration++
+		return zattest.ErrControllerReqFailed
+	}
+
+	//Reset Iteration on successful reply, irrespective of the reply content
+	attestCtx.Iteration = 0
+
+	attestResp := &attest.ZAttestResponse{}
+	if err := proto.Unmarshal(contents, attestResp); err != nil {
+		log.Errorf("[ATTEST] Error %v in Unmarshaling nonce response", err)
+		return zattest.ErrControllerReqFailed
+	}
+
+	respType := attestResp.GetRespType()
+	if respType != attest.ZAttestRespType_ATTEST_RESP_NONCE {
+		log.Errorf("[ATTEST] Got %v, but want %v",
+			respType, attest.ZAttestRespType_ATTEST_RESP_NONCE)
+		return zattest.ErrControllerReqFailed
+	}
+
+	if nonceResp := attestResp.GetNonce(); nonceResp == nil {
+		log.Errorf("[ATTEST] Got empty nonce response")
+		return zattest.ErrControllerReqFailed
+	} else {
+		attestCtx.Nonce = nonceResp.GetNonce()
+	}
+
+	return nil
+}
+
+func getPCRValues(internalQuote *types.AttestQuote, quoteMsg *attest.ZAttestQuote) error {
+	quoteMsg.PcrValues = make([]*attest.TpmPCRValue, 0)
+	for _, pcr := range internalQuote.PCRs {
+		pcrValue := new(attest.TpmPCRValue)
+		pcrValue.Index = uint32(pcr.Index)
+		switch pcr.Algo {
+		case types.PCRExtendHashAlgoSha1:
+			pcrValue.HashAlgo = attest.TpmHashAlgo_TPM_HASH_ALGO_SHA1
+		case types.PCRExtendHashAlgoSha256:
+			pcrValue.HashAlgo = attest.TpmHashAlgo_TPM_HASH_ALGO_SHA256
+		default:
+			return fmt.Errorf("Unknown Hash Algo in PCR Digest %d",
+				pcr.Index)
+		}
+		pcrValue.Value = pcr.Digest
+		quoteMsg.PcrValues = append(quoteMsg.PcrValues, pcrValue)
+	}
+	//XXX Check for TPM platform, and if so, insist on non-empty quoteMsg.PCRValues
 	return nil
 }
 
 //SendAttestQuote implements SendAttestQuote method of zattest.Verifier
 func (server *VerifierImpl) SendAttestQuote(ctx *zattest.Context) error {
-	//XXX: Fill it in when Controller code is ready
+	if ctx.OpaqueCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to OpaqueCtx")
+	}
+	attestCtx, ok := ctx.OpaqueCtx.(*attestContext)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected type from opaque ctx: %T",
+			ctx.OpaqueCtx)
+	}
+	var attestReq = &attest.ZAttestReq{}
+
+	// bail if V2API is not supported
+	if !zedcloud.UseV2API() {
```

Comment:
Feels like we should have this V2 check at the top somewhere, but given that the V1 code will disappear soon I don't think we need to worry about putting the checks somewhere else.